lib/fizzy/stack.hpp

```diff
@@ -1,38 +1,109 @@
 #pragma once
 
+#include <cassert>
 #include <cstdint>
+#include <memory>
 #include <vector>
 
 namespace fizzy
 {
 template <typename T>
-class Stack : public std::vector<T>
+class Stack : private std::vector<T>
 {
 public:
-    using difference_type = typename std::vector<T>::difference_type;
-
-    using std::vector<T>::vector;
-
-    using std::vector<T>::back;
-    using std::vector<T>::emplace_back;
-    using std::vector<T>::pop_back;
-    using std::vector<T>::resize;
+    using std::vector<T>::clear;
+    using std::vector<T>::empty;
     using std::vector<T>::size;
 
-    // Also used: size(), resize(), clear(), empty(), end()
+    /// Push an item to the stack.
+    void push(T val) { std::vector<T>::emplace_back(val); }
 
-    void push(T val) { emplace_back(val); }
+    /// Returns the reference to the top item on the stack.
+    ///
+    /// This is not currently in use (except unit tests), but left as is classic stack method.
+    T& top() noexcept { return std::vector<T>::back(); }
 
+    /// Pops an item from the top of the stack and returns it.
     T pop()
     {
-        const auto res = back();
-        pop_back();
+        const auto res = top();
+        std::vector<T>::pop_back();
         return res;
     }
 
-    T peek(size_t depth = 0) const noexcept { return (*this)[size() - depth - 1]; }
-
     /// Drops @a num_elements elements from the top of the stack.
-    void drop(size_t num_elements = 1) noexcept { resize(size() - num_elements); }
+    void drop(size_t num_elements) noexcept { std::vector<T>::resize(size() - num_elements); }
+};
+
+class OperandStack
+{
+    /// The pointer to the top item, or below the stack bottom if stack is empty.
+    ///
+    /// This pointer always alias m_storage, but it is kept as the first field
+    /// because it is accessed the most. Therefore, it must be initialized
+    /// in the constructor after the m_storage.
+    uint64_t* m_top;
+
+    /// The storage for items.
+    std::unique_ptr<uint64_t[]> m_storage;
+
+public:
+    /// Default constructor. Sets the top item pointer to below the stack bottom.
+    explicit OperandStack(size_t max_stack_height)
+      : m_storage{std::make_unique<uint64_t[]>(max_stack_height)}
+    {
+        m_top = m_storage.get() - 1;
+    }
+
+    OperandStack(const OperandStack&) = delete;
+    OperandStack& operator=(const OperandStack&) = delete;
+
+    /// The current number of items on the stack (aka stack height).
+    [[nodiscard]] size_t size() const noexcept
+    {
+        return static_cast<size_t>(m_top + 1 - &m_storage[0]);
```

Comment:
Is `&m_storage[0]` fine when you call this with 0 max height size in tests?

Reply:
It looks fine, but I can use `m_storage.get()`.